machine_translation/sampling.py

```diff
@@ -181,10 +180,11 @@ def _evaluate_model(self):
         total_cost = 0.0
 
         # Get target vocabulary
-        if not self.trg_ivocab:
-            sources = self._get_attr_rec(self.main_loop, 'data_stream')
-            trg_vocab = sources.data_streams[1].dataset.dictionary
-            self.trg_ivocab = {v: k for k, v in trg_vocab.items()}
+        sources = self._get_attr_rec(self.main_loop, 'data_stream')
+        trg_vocab = sources.data_streams[1].dataset.dictionary
+        self.trg_ivocab = {v: k for k, v in trg_vocab.items()}
+        trg_eos_sym = sources.data_streams[1].dataset.eos_token
+        self.trg_eos_idx = trg_vocab[trg_eos_sym]
```

Comment:
As explained above, it's better to obtain `trg_ivocab` through `sources.data_streams[1].dataset.dictionary`.